### ISSUE TITLE
Comparison with SDDP.jl

### ISSUE THREAD

**Issue by odow** [zidanessf/RDDP.jl#1]
Hi there! Nice work. 

It seems like you should be able to integrate this into SDDP.jl if you had a function which took the domain of `ξ` and returned a vector of the extreme points. Then your README example would be:
```julia
msro = SDDP.LinearPolicyGraph(
    stages = 10,
    sense = :Min,
    lower_bound = 0.0,
    optimizer = Gurobi.Optimizer,
) do ro, t
    @variable(ro, 0 <= x <= 5, SDDP.State, initial_value = 0)
    @variable(ro, 0 <= u <= 1)
    @variable(ro, ξ[i = 1:2])
    @constraint(ro, x.out == x.in - u + sum(ξ))
    @objective(ro, Min, x)
    SDDP.parameterize(ro, [[0.0, 1.0], [1.0, 0.0]]) do ω
        fix.(ξ, ω; force = true)
    end
end
SDDP.train(msro; risk_measure = SDDP.WorstCase())
```
Is this correct or am I missing something?

**Comment by zidanessf** [comment]
> Hi there! Nice work.
> 
> It seems like you should be able to integrate this into SDDP.jl if you had a function which took the domain of `ξ` and returned a vector of the extreme points. Then your README example would be:
> 
> ```julia
> msro = SDDP.LinearPolicyGraph(
>     stages = 10,
>     sense = :Min,
>     lower_bound = 0.0,
>     optimizer = Gurobi.Optimizer,
> ) do ro, t
>     @variable(ro, 0 <= x <= 5, SDDP.State, initial_value = 0)
>     @variable(ro, 0 <= u <= 1)
>     @variable(ro, ξ[i = 1:2])
>     @constraint(ro, x.out == x.in - u + sum(ξ))
>     @objective(ro, Min, x)
>     SDDP.parameterize(ro, [[0.0, 1.0], [1.0, 0.0]]) do ω
>         fix.(ξ, ω; force = true)
>     end
> end
> SDDP.train(msro; risk_measure = SDDP.WorstCase())
> ```
> 
> Is this correct or am I missing something?

That would be perfect to integrate RDDP into SDDP.jl.  However, there are several problems to be further considered.  Firstly, in your new version of README, the uncertain parameters are preferred to be described by the uncertainty set, not directly by the extreme points. Listing all extreme points is not beautiful. Therefore, the second argument of the SDDP.parameterize() should be a JuMP model that describe the support of the uncertain parameters, or a function that generates this model. Secondly, RDDP seems unable to handle the objective state, belief state and the stage dependent uncertainty.

**Comment by odow** [comment]
Okay, multiple questions:

 * Is the worst-case realization an extreme point of the uncertainty set? 
   * If linear I guess so. I assume the argument is that they all have to be feasible, so maybe also if the uncertainty is just in the right-hand side?
 * You're right that enumerating the extreme points is tedious. There might be a lot of them. (Or infinite if not polyhedral.) Could you do some sort of scenario generation scheme?
   * Start with a single realization.
   * Do a forward pass. 
   * For each node, compute the worst-case noise and add that to the set of possible realizations
   * This would only work for problems with the noise in the right-hand side. 
   * It seems like this should work because dual feasibility is independent of the RHS
   * It's been a while since I last read the RDDP paper, so perhaps this is what is already does

**Comment by odow** [comment]
cc @haoxiangyang89 (Prof. at CUHK-Shenzhen) who's also interested in SDDP + Julia

**Comment by zidanessf** [comment]
> Okay, multiple questions:
> 
> * Is the worst-case realization an extreme point of the uncertainty set?
Yes, and only the RHS uncertainty is considered in this implementation. As for the coefficient uncertainty, it would be plausible if one can make sure the stage problem being bounded for all realization of the uncertainty. 
> * You're right that enumerating the extreme points is tedious. There might be a lot of them. (Or infinite if not polyhedral.) Could you do some sort of scenario generation scheme?
The realization is generated by the upper problem, in which the current guess of the worst-case vertex is calculated at each stage via vertex enumeration. It seems unnecessary to maintain a set of possible realizations, because when forward pass and the backward pass go further, the uncertainty trajectory would converge to the true worst-case scenario.

**Comment by odow** [comment]
Another question: 

I arrive at a stage with `x.in`. How does the ordering of decision happen?

 a) I choose `u`, then the other player chooses `ξ`, and from that we compute `x.out`
 b) The other player sees `x.in` and chooses `ξ`, then I get to pick `u`

Because if it's a), that's hard.

If it's b), I could see us adding a callback where you get to modify the sample space based on the incoming state `x`.

I guess a) and b) are just a modeling choice. You can go from b) to a) by the state-space expansion trick where you choose `u` for the next stage and pass it forward as a state variable.

**Comment by zidanessf** [comment]
> Another question:
> 
> I arrive at a stage with `x.in`. How does the ordering of decision happen?
> 
> a) I choose `u`, then the other player chooses `ξ`, and from that we compute `x.out` b) The other player sees `x.in` and chooses `ξ`, then I get to pick `u`

It's b, so called _hazard-decision_ (I'm not sure about this term) process. This repo is made for economic dispatch of power system under renewable generation uncertainties, so it's naturally modeled as the player responses after he observes the uncertainty outcome.  BTW, I'm a little confused about the callback what you say, could you please explain it further?

**Comment by odow** [comment]
> It's b, so called hazard-decision (I'm not sure about this term) process

Yeah. I think it comes from the French community, so "le hasard" https://fr.wikipedia.org/wiki/Hasard. It also kinda works in English. Some people would also call it wait-and-see

> BTW, I'm a little confused about the callback what you say, could you please explain it further?

I've been thinking about how to add decision-dependent uncertainty a bit recently. This is one extreme. And so "incoming state-dependent uncertainty" is potentially easier to implement.

The syntax could be:
```Julia
msro = SDDP.LinearPolicyGraph(
    stages = 10,
    sense = :Min,
    lower_bound = 0.0,
    optimizer = Gurobi.Optimizer,
) do ro, t
    @variable(ro, 0 <= x <= 5, SDDP.State, initial_value = 0)
    @variable(ro, 0 <= u <= 1)
    @variable(ro, ξ[i = 1:2])
    @constraint(ro, x.out == x.in - u + sum(ξ))
    @objective(ro, Min, x)
    SDDP.parameterize(ro, DecisionDependent()) do x
        # Solve max problem for ω
        fix.(ξ, ω; force = true)
    end
end
SDDP.train(msro; risk_measure = SDDP.WorstCase())
```
